RecoTracker/LSTCore/src/alpaka/Segment.h

```diff
@@ -1007,4 +910,17 @@ namespace ALPAKA_ACCELERATOR_NAMESPACE::lst {
   };
 }  // namespace ALPAKA_ACCELERATOR_NAMESPACE::lst
 
+namespace cms::alpakatools {
+  // This is not used, but it is needed for compilation
+  template <typename T0, typename... Args>
+  struct CopyToHost<PortableHostMultiCollection<T0, Args...>> {
```

Comment:
Could you elaborate? Having to copy a "host collection" to host seems wrong.

Reply:
I wanted to ask you about this. I needed to add it so that this part compiles (even with the `if constexpr`).
https://github.com/SegmentLinking/cmssw/blob/96be4f729e1750923f3fbd5903dd819bbe5fbb73/RecoTracker/LSTCore/src/alpaka/Event.dev.cc#L1399-L1412

Reply:
The `if constexpr` allows the non-taken branch to be non-compilable only when the condition depends on a template argument (e.g. if the the `getSegments()` would take the `Queue` as an argument, but could be achieved in many other ways). I'd want to avoid the specialization of `CopyToHost` for host-specific data types.

Reply:
On a similar note: is it intended to make a copy of the data when running on the host ?

Reply:
In fact, what happens if you avoid the copy at line 1404 when running on the Cpu ?
Does the code require that the device buffer is still valid after the copy ?  Does it assume that the "host" and "device" buffers can be modified independently of each other ?

Reply:
Ah, ignore me, I missed the `if` on line 1401.

Reply:
@makortel I tried this
```cpp
template <typename TSoA, typename TDev = Device>
typename TSoA::ConstView Event::getSegments(bool sync) {
  if constexpr (std::is_same_v<TDev, DevHost>)
    return segmentsDev_->const_view<TSoA>();
  if (!segmentsHost_) {
    segmentsHost_.emplace(cms::alpakatools::CopyToHost<SegmentsDeviceCollection>::copyAsync(queue_, *segmentsDev_));
    if (sync)
      alpaka::wait(queue_);  // host consumers expect filled data
  }
  return segmentsHost_->const_view<TSoA>();
}
```

and this
```cpp
template <typename TSoA, typename TQueue>
typename TSoA::ConstView Event::getSegments(bool sync, TQueue& queue) {
  if constexpr (std::is_same_v<alpaka::Dev<TQueue>, DevHost>)
    return segmentsDev_->const_view<TSoA>();
  if (!segmentsHost_) {
    segmentsHost_.emplace(cms::alpakatools::CopyToHost<SegmentsDeviceCollection>::copyAsync(queue_, *segmentsDev_));
    if (sync)
      alpaka::wait(queue_);  // host consumers expect filled data
  }
  return segmentsHost_->const_view<TSoA>();
}
```
but none of them compile.

Reply:
It seems you need to put the non-compilable code into an explicit `if` or `else` branch, along
```cpp
template <typename TSoA, typename TQueue>
typename TSoA::ConstView Event::getSegments(bool sync, TQueue& queue) {
  if constexpr (std::is_same_v<alpaka::Dev<TQueue>, DevHost>) {
    return segmentsDev_->const_view<TSoA>();
  } else {
    if (!segmentsHost_) {
      segmentsHost_.emplace(cms::alpakatools::CopyToHost<SegmentsDeviceCollection>::copyAsync(queue_, *segmentsDev_));
      if (sync)
        alpaka::wait(queue_);  // host consumers expect filled data
    }
    return segmentsHost_->const_view<TSoA>();
  }
}
```

Reply:
Thank you @makortel. For some reason that still didn't work for me, but I got it to work by explicitly involving the templated type in CopyToHost.